src/Compiler/Checking/TailCallChecks.fs

```diff
@@ -0,0 +1,868 @@
+// Copyright (c) Microsoft Corporation.  All Rights Reserved.  See License.txt in the project root for license information.
+
+module internal FSharp.Compiler.TailCallChecks
+
+open Internal.Utilities.Collections
+open Internal.Utilities.Library
+open Internal.Utilities.Library.Extras
+open FSharp.Compiler
+open FSharp.Compiler.AbstractIL.IL
+open FSharp.Compiler.DiagnosticsLogger
+open FSharp.Compiler.Features
+open FSharp.Compiler.Syntax
+open FSharp.Compiler.TcGlobals
+open FSharp.Compiler.TypedTree
+open FSharp.Compiler.TypedTreeBasics
+open FSharp.Compiler.TypedTreeOps
+open FSharp.Compiler.TypeRelations
+
+let PostInferenceChecksStackGuardDepth = GetEnvInteger "FSHARP_TailCallChecks" 50
+
+//--------------------------------------------------------------------------
+// check environment
+//--------------------------------------------------------------------------
+
+type env =
+    {
+        /// Values in module that have been marked [<TailCall>]
+        mustTailCall: Zset<Val>
+
+        /// Recursive expressions of [<TailCall>] attributed values
+        mustTailCallExprs: Map<Stamp, Expr>
+    }
+
+    override _.ToString() = "<env>"
+
+let (|ValUseAtApp|_|) e =
+    match e with
+    | InnerExprPat (Expr.App(funcExpr = InnerExprPat (Expr.Val (valRef = vref; flags = valUseFlags))) | Expr.Val (valRef = vref
+                                                                                                                  flags = valUseFlags)) ->
+        Some(vref, valUseFlags)
+    | _ -> None
+
+type IsTailCall =
+    | Yes of bool // true indicates "has unit return type and must return void"
```

Comment:
IMO this is not expressive enough. I'd prefer having a separate type for this one, smth along the lines of `type = TailCallReturnType = | Void | NonVoid` or `type IsVoid = | Yes | No`.

But, from the design perspective - why is this supposed to matter and not be an implementation detail? We don't distinguish between "void" and "non-void" that much so I wonder if this is an anyhow common pattern in our code. In other words, why cannot this just carry the return type, be it unit or anything else?

Reply:
Oh yeah, that's a lot nicer.

The `void` vs `non-void` handling comes from the checks in [IlxGen.fs](https://github.com/dotnet/fsharp/blob/723b1dc6310d773070804c0d258b2227c77d1ef8/src/Compiler/CodeGen/IlxGen.fs#L4389) when deciding if we can tailcall.
But I'll do some testing to make sure, it's really needed for our purposes and maybe add a test case for it.

Reply:
Oh I see. Alright, if you feel like it's better to align with that thing, it's fine.